### DISCUSSION TITLE
How to get asset name from Handle?

### DISCUSSION THREAD

**Discussion by Moanrisy** [bevyengine/bevy#9006] [Q&A] [answered]
So I'm using Query to get Handle<Image>. Then I want to get assets name from Handle that I loop.
I want to apply transform.rotation to specific assets called "male_arm.png", but I only know the handle.

But I don't know how to proceed from there. 

in handle.rs, it say
To get the actual asset, try using
/// [`Assets::get`]

But I don't know how to use that.
Assets::get(&self, handle)

the &self give me an error "expected value, found module `self`
`self` value is a keyword only available in methods with a `self` parameter"

But I don't know what value to pass in Assets::get first argument, what is the value should?

**Comment by MrGVSV** [comment]
> But I don't know how to use that.
> Assets::get(&self, handle)

You actually need to add `Res<Assets<MyAsset>>` to your system. Then you can take it and call `assets.get(&handle)`.

> But I don't know what value to pass in Assets::get first argument, what is the value should?

`self` is a reserved keyword in Rust. If you haven't already, I recommend going through the [Rust book](https://doc.rust-lang.org/book/).

> I want to apply transform.rotation to specific assets called "male_arm.png", but I only know the handle.

For an `Image` asset, you could maybe get the path using [`AssetServer::get_handle_path`](https://docs.rs/bevy/latest/bevy/asset/struct.AssetServer.html#method.get_handle_path). However, that doesn't always work. Another solution would be to add a component that identifies the specific asset(s) you want to target, then use that to find which entity to snag your handle from.

Lastly, you don't apply a transform to an asset. You do that to an _entity_ which holds a handle to an asset. I think that's what you meant, but I just wanted to make that clear.

**Comment by Moanrisy** [comment]
> You actually need to add `Res<Assets<MyAsset>>` to your system. Then you can take it and call `assets.get(&handle)`.

How to do that? I thought .add_system only for function.
 
> But I don't know what value to pass in Assets::get first argument, what is the value should?

I see, so if the first param &self (lower case S), it mean the instance which called. So I can't use Assets::get() directly. I need to use an instance to call get right?

that's why you use 'assets.get(&handle)'
because assets is the instance and the second param filled with first argument.

> For an `Image` asset, you could maybe get the path using [`AssetServer::get_handle_path`](https://docs.rs/bevy/latest/bevy/asset/struct.AssetServer.html#method.get_handle_path). However, that doesn't always work. Another solution would be to add a component that identifies the specific asset(s) you want to target, then use that to find which entity to snag your handle from. 
> Lastly, you don't apply a transform to an asset. You do that to an _entity_ which holds a handle to an asset. I think that's what you meant, but I just wanted to make that clear.

I think I got confused in here because I don't really get what component,entity means.


thanks for the reply, I will try to re read the concept again before jump into the code.

**Reply by MrGVSV** [reply, marked as answer]
> > You actually need to add `Res<Assets<MyAsset>>` to your system. Then you can take it and call `assets.get(&handle)`.
> 
> How to do that? I thought .add_system only for function.

I mean add it to your function params:

```rust
fn my_system(query: Query<(&mut Transform, &Handle<Image>)>, assets: Res<Assets<Image>>) {
  // ...
}
```

> > But I don't know what value to pass in Assets::get first argument, what is the value should?
> 
> I see, so if the first param &self (lower case S), it mean the instance which called. So I can't use Assets::get() directly. I need to use an instance to call get right?
> 
> that's why you use 'assets.get(&handle)' because assets is the instance and the second param filled with first argument.

Yep, that's right! You need an instance of `Assets<T>`.

> > For an `Image` asset, you could maybe get the path using [`AssetServer::get_handle_path`](https://docs.rs/bevy/latest/bevy/asset/struct.AssetServer.html#method.get_handle_path). However, that doesn't always work. Another solution would be to add a component that identifies the specific asset(s) you want to target, then use that to find which entity to snag your handle from.
> > Lastly, you don't apply a transform to an asset. You do that to an _entity_ which holds a handle to an asset. I think that's what you meant, but I just wanted to make that clear.
> 
> I think I got confused in here because I don't really get what component,entity means.
> 
> thanks for the reply, I will try to re read the concept again before jump into the code.

That's okay! ECS takes a bit of time to get used to, especially if you come from a more OOP background. I would say definitely go through the Rust book (doesn't have to be super in-depth), and then check out Bevy's own [learning materials](https://bevyengine.org/learn/) as well as the [Unofficial Cheatbook](https://bevy-cheatbook.github.io/).

**Reply by Moanrisy** [reply]
Finally it working :) 
![rotate_hand](https://github.com/bevyengine/bevy/assets/17421945/2bedbabe-e4dc-47db-8c55-6519147d1802)

The problem solved after I know how to use Query. I just need to add Arm component on tuple where I put SpriteBundle.
Because SpriteBundle provide Transform, I can query it. But this will select all entities that have Transform component.
So I need filter Transform Query with With<Arm>, this will only select Transform component from entity that have Arm component. 
```
fn animate_character_parts(
    // parents_query: Query<(Entity, &Children), With<Body>>,
    parents_query: Query<(&Name, &Children), With<Character>>,
    mut transform_query: Query<&mut Transform, With<Arm>>,
    time: Res<Time>,
) {
    for (name, children) in parents_query.iter() {
        println!("name is: {:?}", name);

        for child in children {
            if let Ok(mut transform) = transform_query.get_mut(*child) {
                let speed = 4.0; // Adjust the speed of the arm swing
                let amplitude = 0.8; // Adjust the amplitude of the arm swing
                let elapsed_time = time.elapsed_seconds();
                let angle_in_radians = amplitude * (speed * elapsed_time).sin();

                transform.rotation = Quat::from_rotation_z(angle_in_radians);

                let shoulder_x_offset = 0.0;
                let shoulder_y_offset = 25.0;

                transform.translation = Vec3::new(shoulder_x_offset, shoulder_y_offset, 0.0);
            }
        }
    }
}
```
